File: sdk/kotlin/idenqa/src/main/kotlin/dev/idenqa/sdk/CaptureSequence.kt
package dev.idenqa.sdk

import java.time.Instant
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive

data class CaptureEvidenceSequence(val sequenceDigest: String, val index: Int, val count: Int,
    val challengeId: String, val capturedAt: Instant, val previousDigest: String? = null) {
    internal fun wire(): Map<String,Any> = linkedMapOf<String,Any>("sequence_digest" to sequenceDigest,
        "index" to index,"count" to count,"challenge_id" to challengeId,"captured_at" to capturedAt.toString()).apply {
        previousDigest?.let { put("previous_digest",it) }
    }
}

suspend fun CaptureJourney.submitSequence(taskId: String, requirement: CaptureRequirement, frames: List<AcquiredFrame>): CaptureJourneySnapshot {
    val snapshot=currentSnapshot()
    val task=snapshot.currentTask ?: throw IdenqaException.NotStarted
    if(task.id!=taskId || task.evidenceType!=requirement.evidenceType || task.artefact!=requirement.artefact ||
        frames.size !in 2..8 || frames.map { it.challengeId } != requirement.challenges.map { it.id }) throw IdenqaException.InvalidConfiguration
    val digests=frames.map { frame -> frame.artifact.bytes().let { bytes -> try { "sha256:"+JourneyJson.sha256Hex(bytes) } finally { bytes.fill(0) } } }
    val manifest=NoticeJson.encode(linkedMapOf("session_id" to (snapshot.verificationId ?: ""),"requirement_id" to requirement.id,
        "challenges" to requirement.challenges.map { it.id }, "digests" to digests))
    val digest="sha256:"+JourneyJson.sha256Hex(manifest)
    frames.forEachIndexed { index, frame ->
        currentCoroutineContext().ensureActive()
        if(frame.quality.failures(requirement.quality).isNotEmpty()) throw IdenqaException.CaptureQuality
        submit(taskId,frame.artifact,"idenqa.method.live_camera",CaptureEvidenceSequence(digest,index,frames.size,
            requirement.challenges[index].id,frame.capturedAt,if(index==0) null else digests[index-1]))
    }
    return refresh()
}

File: sdk/swift/Sources/Idenqa/CaptureSequence.swift
import Foundation

public struct CaptureEvidenceSequence: Codable, Equatable, Sendable {
    public let sequenceDigest: String
    public let index: Int
    public let count: Int
    public let challengeID: String
    public let capturedAt: Date
    public let previousDigest: String?
    enum CodingKeys: String, CodingKey {
        case index, count
        case sequenceDigest = "sequence_digest", challengeID = "challenge_id", capturedAt = "captured_at", previousDigest = "previous_digest"
    }
}

extension CaptureJourney {
    /// Submits the complete measured sequence in order; Core alone confirms the task.
    public func submitSequence(taskID: String, requirement: CaptureRequirement, frames: [AcquiredFrame]) async throws -> CaptureJourneySnapshot {
        let snapshot = currentSnapshot()
        guard let task = snapshot.currentTask, task.id == taskID,
              task.evidenceType == requirement.evidenceType, task.artefact == requirement.artefact,
              (2...8).contains(frames.count), frames.count == requirement.challenges.count,
              frames.map(\.challengeID) == requirement.challenges.map({ Optional($0.id) }) else { throw IdenqaError.invalidConfiguration }
        let digests = frames.map { "sha256:" + CaptureDigest.sha256Hex($0.artifact.bytes) }
        let manifest = try JSONSerialization.data(withJSONObject: ["session_id": snapshot.verificationID ?? "", "requirement_id": requirement.id,
            "challenges": requirement.challenges.map(\.id), "digests": digests], options: [.sortedKeys])
        let sequenceDigest = "sha256:" + CaptureDigest.sha256Hex(manifest)
        for (index, frame) in frames.enumerated() {
            try Task.checkCancellation()
            guard frame.quality.failures(against: requirement.quality).isEmpty else { throw IdenqaError.captureQuality }
            _ = try await submit(taskID: taskID, artifact: frame.artifact, method: "idenqa.method.live_camera",
                sequence: CaptureEvidenceSequence(sequenceDigest: sequenceDigest, index: index, count: frames.count,
                    challengeID: requirement.challenges[index].id, capturedAt: frame.capturedAt,
                    previousDigest: index == 0 ? nil : digests[index - 1]))
        }
        return try await refresh()
    }
}
